clair/src/main/scala-3/Macros.scala

```diff
@@ -450,63 +450,80 @@ def partitionedConstructs[Def <: OpInputDef: Type](
     tpe match
-      case '[t] =>
+      case '[type t <: Attribute; `t`] =>
         variadicity match
           case Variadicity.Optional =>
             // If the construct is optional, check if it is defined and if so, verify its type
             '{
-              if (!${ c }.isInstanceOf[Seq[DefinedInputOf[Def, t & Attribute]]])
+              if (!${ c }.isInstanceOf[Seq[DefinedInputOf[Def, t]]])
               then
                 throw new Exception(
                   s"Expected ${${ Expr(d.name) }} to be of type ${${
-                      Expr(Type.show[DefinedInputOf[Def, t & Attribute]])
+                      Expr(Type.show[DefinedInputOf[Def, t]])
                     }}, got ${${ c }}"
                 )
               ${ c }
-                .asInstanceOf[Seq[DefinedInputOf[Def, t & Attribute]]]
+                .asInstanceOf[Seq[DefinedInputOf[Def, t]]]
                 .headOption
             }
           // If the construct is not variadic, just check if it is of the expected type
           case Variadicity.Single =>
             '{
-              if (!${ c }.isInstanceOf[DefinedInputOf[Def, t & Attribute]]) then
+              if (!${ c }.isInstanceOf[DefinedInputOf[Def, t]]) then
                 throw new Exception(
                   s"Expected ${${ Expr(d.name) }} to be of type ${${
-                      Expr(Type.show[DefinedInputOf[Def, t & Attribute]])
+                      Expr(Type.show[DefinedInputOf[Def, t]])
                     }}, got ${${ c }}"
                 )
 
-              ${ c }.asInstanceOf[DefinedInputOf[Def, t & Attribute]]
+              ${ c }.asInstanceOf[DefinedInputOf[Def, t]]
             }
           // If the construct is variadic, check if it is a list of the expected type
           case Variadicity.Variadic =>
             '{
               if (
                 !${ c }
-                  .isInstanceOf[Seq[DefinedInputOf[Def, t & Attribute]]]
+                  .isInstanceOf[Seq[DefinedInputOf[Def, t]]]
               ) then
                 throw new Exception(
                   s"Expected ${${ Expr(d.name) }} to be of type ${${
-                      Expr(Type.show[Seq[DefinedInputOf[Def, t & Attribute]]])
+                      Expr(Type.show[Seq[DefinedInputOf[Def, t]]])
                     }}, got ${${ c }}"
                 )
               ${ c }
-                .asInstanceOf[Seq[DefinedInputOf[Def, t & Attribute]]]
+                .asInstanceOf[Seq[DefinedInputOf[Def, t]]]
                 .toSeq
             }
   }
 }
 
+def verifiedConstructs[Def <: OpInputDef: Type](
+    defs: Seq[Def],
+    op: Expr[DerivedOperationCompanion[?]#UnverifiedOp]
+)(using Quotes) = {
+  // Get the flat sequence of these constructs
+  val flat = getConstructSeq(op)
+  // partition the constructs according to their definitions
+  val partitioned = partitionConstructs(
+    defs,
+    op,
+    flat
+  )
+
+  // Verify the constructs
+  verifiyConstructs(defs, op, partitioned)
+}
```

Comment:
There are a few polishes around, but this is the main idea hear; I'm trying to decouple different steps for whice we generate functions, to help potentially reusing some in other parts of the framework (i.e., declarative assembly format)